frmts/grok/grokdataset.cpp

```diff
@@ -0,0 +1,4173 @@
+        if( bUseXLBoxes )
+        {
+            GUIntBig nXLBox = 0;
+            VSIFWriteL(&nXLBox, 1, 8, fp);
+        }
+    }
+
+/* -------------------------------------------------------------------- */
+/*      Do lossless reuse of an existing JPEG2000 codestream            */
+/* -------------------------------------------------------------------- */
+    if( fpSrc )
+    {
+        const char* apszIgnoredOptions[] = {
+            "BLOCKXSIZE", "BLOCKYSIZE", "QUALITY", "REVERSIBLE",
+            "RESOLUTIONS", "PROGRESSION", "SOP", "EPH",
+            "YCBCR420", "YCC", "NBITS", "1BIT_ALPHA", "PRECINCTS",
+            "TILEPARTS", "CODEBLOCK_WIDTH", "CODEBLOCK_HEIGHT", "PLT", nullptr };
+        for( int i = 0; apszIgnoredOptions[i]; i ++)
+        {
+            if( CSLFetchNameValue(papszOptions, apszIgnoredOptions[i]) )
+            {
+                CPLError(CE_Warning, CPLE_NotSupported,
+                            "Option %s ignored when USE_SRC_CODESTREAM=YES",
+                            apszIgnoredOptions[i]);
+            }
+        }
+        GByte abyBuffer[4096];
+        VSIFSeekL( fpSrc, nCodeStreamStart, SEEK_SET );
+        vsi_l_offset nRead = 0;
+        /* coverity[tainted_data] */
+        while( nRead < nCodeStreamLength )
+        {
+            int nToRead = ( nCodeStreamLength-nRead > 4096 ) ? 4096 :
+                                        (int)(nCodeStreamLength-nRead);
+            if( (int)VSIFReadL(abyBuffer, 1, nToRead, fpSrc) != nToRead )
+            {
+                VSIFCloseL(fp);
+                VSIFCloseL(fpSrc);
+                grk_object_unref(&psImage->obj);
+                grk_object_unref(pCodec);
+                delete poGMLJP2Box;
+                return nullptr;
+            }
+            if( nRead == 0 && (pszProfile || bInspireTG) &&
+                abyBuffer[2] == 0xFF && abyBuffer[3] == 0x51 )
+            {
+                if( EQUAL(pszProfile, "UNRESTRICTED") )
+                {
+                    abyBuffer[6] = 0;
+                    abyBuffer[7] = 0;
+                }
+                else if( EQUAL(pszProfile, "PROFILE_1") || bInspireTG )
+                {
+                    // TODO: ultimately we should check that we can really set Profile 1
+                    abyBuffer[6] = 0;
+                    abyBuffer[7] = 2;
+                }
+            }
+            if( (int)VSIFWriteL(abyBuffer, 1, nToRead, fp) != nToRead ||
+                !pfnProgress( (nRead + nToRead) * 1.0 / nCodeStreamLength,
+                                nullptr, pProgressData ) )
+            {
+                VSIFCloseL(fp);
+                VSIFCloseL(fpSrc);
+                grk_object_unref(&psImage->obj);
+                grk_object_unref(pCodec);
+                delete poGMLJP2Box;
+                return nullptr;
+            }
+            nRead += nToRead;
+        }
+
+        VSIFCloseL(fpSrc);
+    }
+    else
+    {
+        JP2GrokFile __sJP2GrokFile;
```

Comment:
https://en.cppreference.com/w/cpp/language/identifiers: the identifiers with a double underscore anywhere are reserved;